From b34a4e2f812e916f806f61412dc3f70f31a32fd0 Mon Sep 17 00:00:00 2001
From: Hin-Tak Leung <htl10@users.sourceforge.net>
Date: Fri, 27 Dec 2024 02:08:12 +0000
Subject: [PATCH] Fix for -Werror=redundant-move for newer g++

Signed-off-by: Hin-Tak Leung <htl10@users.sourceforge.net>
---
 src/gn/desc_builder.cc | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/gn/desc_builder.cc b/src/gn/desc_builder.cc
index 444a5e02..3ef170c3 100644
--- a/src/gn/desc_builder.cc
+++ b/src/gn/desc_builder.cc
@@ -167,7 +167,7 @@ class BaseDescBuilder {
     base::ListValue res;
     for (const auto& v : vector)
       res.GetList().emplace_back(ToBaseValue(v));
-    return std::move(res);
+    return res;
   }
 
   base::Value ToBaseValue(const Scope* scope) {
@@ -176,7 +176,7 @@ class BaseDescBuilder {
     scope->GetCurrentScopeValues(&map);
     for (const auto& v : map)
       res.SetKey(v.first, ToBaseValue(v.second));
-    return std::move(res);
+    return res;
   }
 
   base::Value ToBaseValue(const Value& val) {
-- 
2.47.1

