diff --git a/third_party/libjpeg-turbo/BUILD.gn b/third_party/libjpeg-turbo/BUILD.gn
index 4cf94f26c2..08799d8a37 100644
--- a/third_party/libjpeg-turbo/BUILD.gn
+++ b/third_party/libjpeg-turbo/BUILD.gn
@@ -87,25 +87,35 @@ if (skia_use_system_libjpeg_turbo) {
       sources += [
         "../externals/libjpeg-turbo/simd/arm/arm/jsimd.c",
         "../externals/libjpeg-turbo/simd/arm/arm/jsimd_neon.S",
+        "../externals/libjpeg-turbo/simd/arm/common/jccolor-neon.c",
+        "../externals/libjpeg-turbo/simd/arm/common/jcgray-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jcsample-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jdcolor-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jdmerge-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jdsample-neon.c",
+        "../externals/libjpeg-turbo/simd/arm/common/jfdctfst-neon.c",
+        "../externals/libjpeg-turbo/simd/arm/common/jfdctint-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jidctfst-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jidctint-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jidctred-neon.c",
+        "../externals/libjpeg-turbo/simd/arm/common/jquanti-neon.c",
       ]
     } else if (current_cpu == "arm64" && !is_ios && !is_win) {
       sources += [
         "../externals/libjpeg-turbo/simd/arm/arm64/jsimd.c",
         "../externals/libjpeg-turbo/simd/arm/arm64/jsimd_neon.S",
+        "../externals/libjpeg-turbo/simd/arm/common/jccolor-neon.c",
+        "../externals/libjpeg-turbo/simd/arm/common/jcgray-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jcsample-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jdcolor-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jdmerge-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jdsample-neon.c",
+        "../externals/libjpeg-turbo/simd/arm/common/jfdctfst-neon.c",
+        "../externals/libjpeg-turbo/simd/arm/common/jfdctint-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jidctfst-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jidctint-neon.c",
         "../externals/libjpeg-turbo/simd/arm/common/jidctred-neon.c",
+        "../externals/libjpeg-turbo/simd/arm/common/jquanti-neon.c",
       ]
     } else {
       sources += [ "../externals/libjpeg-turbo/jsimd_none.c" ]
